67_RayQueryGeometry/main.cpp

```diff
@@ -0,0 +1,1088 @@
+			objectsCpu[OT_CONE] = ReferenceObjectCpu{ .meta = {.type = OT_CONE, .name = "Cone Mesh" }, .shadersType = GP_CONE, .data = gc->createConeMesh(2, 3, 10) };
+			objectsCpu[OT_ICOSPHERE] = ReferenceObjectCpu{ .meta = {.type = OT_ICOSPHERE, .name = "Icosphere Mesh" }, .shadersType = GP_ICO, .data = gc->createIcoSphere(1, 3, true) };
+
+			auto cmdbuf = getSingleUseCommandBufferAndBegin(pool);
+
+			for (uint32_t i = 0; i < objectsCpu.size(); i++)
+			{
+				const auto& geom = objectsCpu[i];
+				auto& obj = objectsGpu[i];
+
+				obj.meta.name = geom.meta.name;
+				obj.meta.type = geom.meta.type;
+
+				obj.indexCount = geom.data.indexCount;
+				obj.indexType = geom.data.indexType;
+
+// TODO: use asset converter to convert all buffers to IGPUBuffers (that part 100% works)
+
+				auto vBuffer = smart_refctd_ptr(geom.data.bindings[0].buffer); // no offset
+				auto vUsage = bitflag(asset::IBuffer::EUF_STORAGE_BUFFER_BIT) | asset::IBuffer::EUF_TRANSFER_DST_BIT | asset::IBuffer::EUF_INLINE_UPDATE_VIA_CMDBUF | 
+					asset::IBuffer::EUF_ACCELERATION_STRUCTURE_BUILD_INPUT_READ_ONLY_BIT | asset::IBuffer::EUF_SHADER_DEVICE_ADDRESS_BIT;
+				obj.bindings.vertex.offset = 0u;
+				auto vertexBuffer = m_device->createBuffer(IGPUBuffer::SCreationParams({.size = vBuffer->getSize(), .usage = vUsage}));
+
+				//if (!vertexBuffer)
+				//	return false;
+
+				auto iBuffer = smart_refctd_ptr(geom.data.indexBuffer.buffer); // no offset
+				auto iUsage = bitflag(asset::IBuffer::EUF_STORAGE_BUFFER_BIT) | asset::IBuffer::EUF_TRANSFER_DST_BIT | asset::IBuffer::EUF_INLINE_UPDATE_VIA_CMDBUF |
+					asset::IBuffer::EUF_ACCELERATION_STRUCTURE_BUILD_INPUT_READ_ONLY_BIT | asset::IBuffer::EUF_SHADER_DEVICE_ADDRESS_BIT;
+				obj.bindings.index.offset = 0u;
+				auto indexBuffer = iBuffer ? m_device->createBuffer(IGPUBuffer::SCreationParams({ .size = iBuffer->getSize(), .usage = iUsage })) : nullptr;
+
+				//if (geom.data.indexType != EIT_UNKNOWN)
+				//	if (!indexBuffer)
+				//		return false;
+
+				for (auto buf : { vertexBuffer, indexBuffer })
+				{
+					if (buf)
+					{
+						auto reqs = buf->getMemoryReqs();
+						reqs.memoryTypeBits &= m_device->getPhysicalDevice()->getUpStreamingMemoryTypeBits();
+						m_device->allocate(reqs, buf.get(), IDeviceMemoryAllocation::EMAF_DEVICE_ADDRESS_BIT);
+					}
+				}
+
+				obj.bindings.vertex = { .offset = 0u, .buffer = std::move(vertexBuffer) };
+				SBufferRange<IGPUBuffer> vRange = { .offset = obj.bindings.vertex.offset, .size = obj.bindings.vertex.buffer->getSize(), .buffer = obj.bindings.vertex.buffer };
+				cmdbuf->updateBuffer(vRange, vBuffer->getPointer());
+
+				if (iBuffer)
+				{
+					obj.bindings.index = { .offset = 0u, .buffer = std::move(indexBuffer) };
+					SBufferRange<IGPUBuffer> iRange = { .offset = obj.bindings.index.offset, .size = obj.bindings.index.buffer->getSize(), .buffer = obj.bindings.index.buffer };
+					cmdbuf->updateBuffer(iRange, iBuffer->getPointer());
+				}
```

Comment:
geometry creator may spit out a mesh thats too big to update a buffer for like that (e.g. more than 1200 vertices), use `IUtilities` to create or update the buffers with data instead (or asset converter!)

There's a 64kb limit to `cmdbuf->updateBuffer`